src/connectors/numeric-menu/connectNumericMenu.ts

```diff
@@ -4,146 +4,157 @@ import {
+  /**
+   * Function to transform the items passed to the templates
+   */
+  transformItems?: ParamTransformItems;
+};
+
+type Refine = (facetValue: string) => void;
+
+export interface NumericMenuRendererOptions<TNumericMenuWidgetParams>
+  extends RendererOptions<TNumericMenuWidgetParams> {
+  /**
+   * The list of available choices
+   */
+  items: Item[];
+  /**
+   * Creates URLs for the next state, the string is the name of the selected option
+   */
+  createURL: (value: Item['value']) => string;
+  /**
+   * `true` if the last search contains no result
+   */
+  hasNoResults: boolean;
+  /**
+   * Sets the selected value and trigger a new search
+   */
+  refine: Refine;
+}
+
+export type NumericMenuRenderer<TNumericMenuWidgetParams> = Renderer<
+  NumericMenuRendererOptions<
+    NumericMenuConnectorParams & TNumericMenuWidgetParams
+  >
+>;
+
+export type NumericMenuWidgetFactory<TNumericMenuWidgetParams> = WidgetFactory<
+  NumericMenuConnectorParams & TNumericMenuWidgetParams
+>;
+
+type NumericMenuConnector = <TNumericMenuWidgetParams>(
+  render: NumericMenuRenderer<TNumericMenuWidgetParams>,
+  unmount?: () => void
+) => NumericMenuWidgetFactory<TNumericMenuWidgetParams>;
+
+const connectNumericMenu: NumericMenuConnector = (
+  renderFn,
+  unmountFn = noop
+) => {
   checkRendering(renderFn, withUsage());
 
-  return (widgetParams = {}) => {
-    const { attribute, items, transformItems = x => x } = widgetParams;
+  return widgetParams => {
+    const {
+      attribute = '',
+      items = [],
+      transformItems = (x => x) as ParamTransformItems,
+    } = widgetParams || {};
 
-    if (!attribute) {
+    if (attribute === '') {
       throw new Error(withUsage('The `attribute` option is required.'));
     }
 
-    if (!items) {
+    if (!items || items.length === 0) {
       throw new Error(
         withUsage('The `items` option expects an array of objects.')
       );
     }
 
+    type ConnectorState = {
+      refine?: Refine;
+      createURL?: (state: SearchParameters) => (facetValue: string) => string;
+      prepareItems?: (state: SearchParameters) => Item[];
+    };
+
+    const connectorState: ConnectorState = {};
+
     return {
       $$type: 'ais.numericMenu',
 
       init({ helper, createURL, instantSearchInstance }) {
-        this._refine = facetValue => {
-          const refinedState = refine(
-            helper.state,
-            attribute,
-            items,
-            facetValue
-          );
+        connectorState.refine = facetValue => {
+          const refinedState = refine(helper.state, attribute, facetValue);
           helper.setState(refinedState).search();
         };
 
-        this._createURL = state => facetValue =>
-          createURL(refine(state, attribute, items, facetValue));
+        connectorState.createURL = state => facetValue =>
+          createURL(refine(state, attribute, facetValue));
 
-        this._prepareItems = state =>
+        connectorState.prepareItems = state =>
```

Comment:
Why is this stored on the connector state? It's a util function so we can pass both `state` and `items` as arguments and remove it from the connector scope.